### DISCUSSION TITLE
google spreadsheets to and from supabase

### DISCUSSION THREAD

**Discussion by davidtew** [supabase/supabase#878] [Questions]
I am hitting speed and size limits with google apps script / google sheets, particularly handling large sheets or complex workflows where sheets are made on the fly to user data in another sheet.  I would love to keep sheets as our interface at work, at least for now, since our organisation knows how to use them, but try the power of postgres to see if I can tackle  these problems.

I have used google apps script a lot (which is essentially javascript) but I have no formal training in programming (so I don't know what I need to know).

I am prepared to embrace it all, but would rather know if its all highly likely to work, technically, before I pour time into it.  (I haven't tried supabase in the wild - that will be a lot of time/work for me so I need to justify it. As of now I am judging supabase by its choice of open source elements woven together, and adopters' reactions)

The spreadsheets data and structure are available by REST via Google Sheets API v4 or, in google apps script only, Class SpreadsheetApp.

I also would put google docs into supabase, again available by REST via  Google Docs API v1

I would need to continue to work in the google apps script environment since its the only one I know.  Should I somehow get the supabase js clients working in google apps script?  Or should I use google apps script Class UrlFetchApp within google apps script and send / receive to / from supabase with a POST request in JSON? 

Are there examples available of any of this already done for me to copy and hack, which is my way!

Has anyone proven that this works?  

Many thanks in advance.

**Comment by mike-wax** [comment]
We started Wax out of a Supabase hackathon. The app that lets you create admin portals / internal tools for your Supabase project in Google Sheets. For example, this demo shows a realtime two way sync between Supabase and Sheets:

https://twitter.com/waxdotrun/status/1447528149579968516

**Comment by Rubenburdin** [comment]
Hey davidtew, coming back to this in 2025 since the Sheets to Supabase question still comes up often.
The good news is that what you described is definitely achievable. The approach depends on how real-time you need things. For occasional syncs, using UrlFetchApp with Supabase's REST API works fine. You authenticate with your anon or service role key, then POST/GET JSON to your tables. The Supabase JS client wasn't really designed for Google Apps Script's runtime, so the REST route is cleaner. There's a decent example in the Supabase docs on using the REST API directly (https://supabase.com/docs/guides/api).
For the two-way sync part, that's where things get tricky. Sheets doesn't have native webhooks, so you end up polling or using onEdit triggers that only fire when users manually change cells. If you're dealing with large datasets and need reliable bi-directional sync without building all the plumbing yourself, [Stacksync](https://www.stacksync.com/) handles exactly this pattern. It keeps your Postgres tables in sync with external systems and propagates changes both directions automatically. Might save you from reinventing the CDC layer in Apps Script.

**Reply by quangle-lab** [reply]
Hey man, thanks for the heads up. I'm also migrating from my AppScripts to a dedicated BE

**Comment by MakeMeDizzy** [comment]
@Rubenburdin If you're going the Apps Script route: the cleanest setup is UrlFetchApp → Supabase REST API (PostgREST) with your service-role key, a time-driven trigger to pull DB changes into the sheet, and an onEdit trigger to push edits back up. That handles occasional syncs fine.

Where it gets painful is making it reliably two-way: Sheets has no real webhooks so you end up polling, onEdit doesn't fire on programmatic/bulk changes, and you have to hand-roll conflict handling (which side wins when the same row changed in both places?) plus type mapping for json/timestamptz. Babysitting that script is what eventually pushed me off the DIY approach.

I'm building a small tool that does exactly this part — two-way Supabase ↔ Sheets with per-table conflict rules (db-wins / sheet-wins / newest-wins) and safe type mapping, so you don't maintain a script. Early access if useful: https://glistening-palmier-d44c63.netlify.app/?utm_source=gh_878
Happy to drop a working Apps Script snippet here if you'd rather keep it DIY.
